sdks/java/harness/src/test/java/org/apache/beam/fn/harness/state/OrderedListUserStateTest.java

```diff
@@ -0,0 +1,597 @@
+            "instructionId",
+            createOrderedListStateKey("A"),
+            StringUtf8Coder.of());
+    assertThat(userState.read(), is(emptyIterable()));
+  }
+
+  @Test
+  public void testRead() throws Exception {
+    FakeBeamFnStateClient fakeClient =
+        new FakeBeamFnStateClient(
+            TimestampedValueCoder.of(StringUtf8Coder.of()),
+            ImmutableMap.of(createOrderedListStateKey("A", 1), asList(A1, B1)));
+    OrderedListUserState<String> userState =
+        new OrderedListUserState<>(
+            Caches.noop(),
+            fakeClient,
+            "instructionId",
+            createOrderedListStateKey("A"),
+            StringUtf8Coder.of());
+
+    assertArrayEquals(
+        asList(A1, B1).toArray(), Iterables.toArray(userState.read(), TimestampedValue.class));
+    userState.asyncClose();
+    assertThrows(IllegalStateException.class, () -> userState.read());
+  }
+
+  @Test
+  public void testReadRange() throws Exception {
+    FakeBeamFnStateClient fakeClient =
+        new FakeBeamFnStateClient(
+            TimestampedValueCoder.of(StringUtf8Coder.of()),
+            ImmutableMap.of(
+                createOrderedListStateKey("A", 1), asList(A1, B1),
+                createOrderedListStateKey("A", 4), Collections.singletonList(A4),
+                createOrderedListStateKey("A", 2), Collections.singletonList(A2)));
+
+    OrderedListUserState<String> userState =
+        new OrderedListUserState<>(
+            Caches.noop(),
+            fakeClient,
+            "instructionId",
+            createOrderedListStateKey("A"),
+            StringUtf8Coder.of());
+
+    Iterable<TimestampedValue<String>> stateBeforeB2 =
+        userState.readRange(Instant.ofEpochMilli(2), Instant.ofEpochMilli(4));
+    assertArrayEquals(
+        Collections.singletonList(A2).toArray(),
+        Iterables.toArray(stateBeforeB2, TimestampedValue.class));
+
+    // Add a new value to an existing sort key
+    userState.add(B2);
+    assertArrayEquals(
+        Collections.singletonList(A2).toArray(),
+        Iterables.toArray(stateBeforeB2, TimestampedValue.class));
+    assertArrayEquals(
+        asList(A2, B2).toArray(),
+        Iterables.toArray(
+            userState.readRange(Instant.ofEpochMilli(2), Instant.ofEpochMilli(4)),
+            TimestampedValue.class));
+
+    // Add a new value to a new sort key
+    userState.add(A3);
+    assertArrayEquals(
+        Collections.singletonList(A2).toArray(),
+        Iterables.toArray(stateBeforeB2, TimestampedValue.class));
+    assertArrayEquals(
+        asList(A2, B2, A3).toArray(),
+        Iterables.toArray(
+            userState.readRange(Instant.ofEpochMilli(2), Instant.ofEpochMilli(4)),
+            TimestampedValue.class));
+
+    userState.asyncClose();
+    assertThrows(
+        IllegalStateException.class,
+        () -> userState.readRange(Instant.ofEpochMilli(1), Instant.ofEpochMilli(2)));
+  }
+
+  @Test
+  public void testAdd() throws Exception {
```

Comment:
Do we need to have some tests where we're partway through an Iterable when the element gets added? In all of the ones I see here we completely process each Iterable before we do any further operations on it (by calling `Iterables.toArray`).

Reply:
Good point! I will add some more tests to cover this.

Reply:
Test added.